mopidy/internal/validation.py

```diff
@@ -1,45 +1,67 @@
+from __future__ import annotations
+
 import urllib.parse
 from collections.abc import Iterable, Mapping
-from typing import Any, Optional, TypeVar, Union
+from typing import Any, Literal, Optional, TypeVar, Union, get_args
 
 from mopidy import exceptions
+from mopidy.audio.constants import PlaybackState
+from mopidy.types import (
+    DistinctField,
+    Query,
+    QueryValue,
+    SearchField,
+    TracklistField,
+)
+
+
+def get_literals(literal_type: Any) -> set[str]:
+    # Check if it's a union
+    if hasattr(literal_type, "__origin__") and literal_type.__origin__ is Union:
+        literals = set()
+        for arg in get_args(literal_type):
+            literals.update(get_literals(arg))
+        return literals
+
+    # Check if it's a literal
+    if hasattr(literal_type, "__origin__") and literal_type.__origin__ is Literal:
+        return set(get_args(literal_type))
+
+    raise ValueError("Provided type is neither a Union nor a Literal type.")
+
 
 T = TypeVar("T")
 
-PLAYBACK_STATES = {"paused", "stopped", "playing"}
+PLAYBACK_STATES: set[str] = {ps.value for ps in PlaybackState}
 
-TRACK_FIELDS_WITH_TYPES: dict[str, Union[type[str], type[int]]] = {
-    "uri": str,
-    "track_name": str,
+FIELD_TYPES: dict[str, type] = {
     "album": str,
-    "artist": str,
     "albumartist": str,
+    "any": Union[int, str],
+    "artist": str,
+    "comment": str,
     "composer": str,
-    "performer": str,
-    "track_no": int,
-    "genre": str,
     "date": str,
-    "comment": str,
     "disc_no": int,
+    "genre": str,
+    "musicbrainz_id": str,
     "musicbrainz_albumid": str,
     "musicbrainz_artistid": str,
     "musicbrainz_trackid": str,
+    "name": str,
+    "performer": str,
+    "tlid": int,
+    "track_name": str,
+    "track_no": int,
+    "uri": str,
```

Comment:
```suggestion
    "uri": Uri,
```
?

Reply:
Not certain about this one. If we change `Uri` from a `TypeAlias` to a `NewType`, we still want a string passed from e.g. the JSON-RPC API to be considered a valid type for the `uri` field in a search query.

Reply:
Could we achieve that with some (pydantic style) type validation and coercion between those layers?

Reply:
> we still want a string passed from e.g. the JSON-RPC API to be considered a valid type

Doesn't this also apply to all other `Uri` parms now it's a NewType? So this will be a breaking change for JSON-RPC clients?

Reply:
I attempted to change `str` to `Uri` here, but then I got a couple of test errors related to logging of backends returning the wrong type from `get_distinct()`. Haven't looked further into it, but I fear this might be a rabbit hole and should be saved for another round of type tightening.

Reply:
Happy with that. Let's leave it and see how we go as it is.